Crash in RemoveFailedPods strategy

<!-- Please answer these questions before submitting your bug report. Thanks! -->

**What version of descheduler are you using?**

descheduler version: 0.22

**Please provide a copy of your descheduler policy config file**
```
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: descheduler-policy-configmap
  namespace: kube-system
data:
  policy.yaml: |
    apiVersion: "descheduler/v1alpha1"
    kind: "DeschedulerPolicy"
    strategies:
      "RemoveFailedPods":
        enabled: true
```
Disabling RemovedFailedPods I can run all other policies.

**What k8s version are you using (`kubectl version`)?**
1.22.1
<details><summary><code>kubectl version</code> Output</summary><br><pre>
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.1", GitCommit:"632ed300f2c34f6d6d15ca4cef3d3c7073412212", GitTreeState:"archive", BuildDate:"2021-08-19T21:28:12Z", GoVersion:"go1.17", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.1", GitCommit:"632ed300f2c34f6d6d15ca4cef3d3c7073412212", GitTreeState:"clean", BuildDate:"2021-08-19T15:39:34Z", GoVersion:"go1.16.7", Compiler:"gc", Platform:"linux/amd64"}

</pre></details>


**What did you do?**

```
I0918 06:44:25.850399       1 named_certificates.go:53] "Loaded SNI cert" index=0 certName="self-signed loopback" certDetail="\"apiserver-loopback-client@1631947465\" [serving] validServingFor=[apiserver-loopback-client] issuer=\"apiserver-loopback-client-ca@1631947464\" (2021-09-18 05:44:23 +0000 UTC to 2022-09-18 05:44:23 +0000 UTC (now=2021-09-18 06:44:25.85028599 +0000 UTC))"
I0918 06:44:25.850606       1 secure_serving.go:195] Serving securely on [::]:10258
I0918 06:44:25.851136       1 tlsconfig.go:240] "Starting DynamicServingCertificateController"
I0918 06:44:25.895398       1 node.go:46] "Node lister returned empty list, now fetch directly"
E0918 06:44:25.919640       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 1 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x14cdf80, 0x26e1770)
	/go/src/sigs.k8s.io/descheduler/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0x84
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/sigs.k8s.io/descheduler/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x80
panic(0x14cdf80, 0x26e1770)
	/usr/local/go/src/runtime/panic.go:965 +0x154
sigs.k8s.io/descheduler/pkg/descheduler/strategies.RemoveFailedPods(0x1a00468, 0x400019c020, 0x1a26ab8, 0x400002ac60, 0x4000255801, 0x0, 0x0, 0x4000602720, 0x6, 0x6, ...)
	/go/src/sigs.k8s.io/descheduler/pkg/descheduler/strategies/failedpods.go:44 +0x70
sigs.k8s.io/descheduler/pkg/descheduler.RunDeschedulerStrategies.func1()
	/go/src/sigs.k8s.io/descheduler/pkg/descheduler/descheduler.go:145 +0x26c
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x4000707ad0)
	/go/src/sigs.k8s.io/descheduler/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x64
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x4000707ad0, 0x19c1ff0, 0x4000454210, 0x1, 0x4000028420)
	/go/src/sigs.k8s.io/descheduler/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0x74
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x4000707ad0, 0x0, 0x0, 0x1, 0x4000028420)
	/go/src/sigs.k8s.io/descheduler/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x88
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go/src/sigs.k8s.io/descheduler/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90
sigs.k8s.io/descheduler/pkg/descheduler.RunDeschedulerStrategies(0x1a00468, 0x400019c020, 0x400046c2c0, 0x40001a6140, 0x40003be487, 0x9, 0x4000028420, 0x4000092c00, 0x40000ee460)
	/go/src/sigs.k8s.io/descheduler/pkg/descheduler/descheduler.go:117 +0x534
sigs.k8s.io/descheduler/pkg/descheduler.Run(0x400046c2c0, 0x19c2050, 0x400021c0e0)
	/go/src/sigs.k8s.io/descheduler/pkg/descheduler/descheduler.go:64 +0x154
sigs.k8s.io/descheduler/cmd/descheduler/app.Run(...)
	/go/src/sigs.k8s.io/descheduler/cmd/descheduler/app/server.go:93
sigs.k8s.io/descheduler/cmd/descheduler/app.NewDeschedulerCommand.func1(0x40003ba280, 0x400012b7c0, 0x0, 0x4)
	/go/src/sigs.k8s.io/descheduler/cmd/descheduler/app/server.go:78 +0xb8
github.com/spf13/cobra.(*Command).execute(0x40003ba280, 0x40001a6010, 0x4, 0x4, 0x40003ba280, 0x40001a6010)
	/go/src/sigs.k8s.io/descheduler/vendor/github.com/spf13/cobra/command.go:856 +0x1d0
github.com/spf13/cobra.(*Command).ExecuteC(0x40003ba280, 0x400063ff48, 0x1, 0x1)
	/go/src/sigs.k8s.io/descheduler/vendor/github.com/spf13/cobra/command.go:960 +0x268
github.com/spf13/cobra.(*Command).Execute(...)
	/go/src/sigs.k8s.io/descheduler/vendor/github.com/spf13/cobra/command.go:897
main.main()
	/go/src/sigs.k8s.io/descheduler/cmd/descheduler/descheduler.go:34 +0xa0
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x12b7020]

goroutine 1 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/sigs.k8s.io/descheduler/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0x108
panic(0x14cdf80, 0x26e1770)
	/usr/local/go/src/runtime/panic.go:965 +0x154
sigs.k8s.io/descheduler/pkg/descheduler/strategies.RemoveFailedPods(0x1a00468, 0x400019c020, 0x1a26ab8, 0x400002ac60, 0x4000255801, 0x0, 0x0, 0x4000602720, 0x6, 0x6, ...)
	/go/src/sigs.k8s.io/descheduler/pkg/descheduler/strategies/failedpods.go:44 +0x70
sigs.k8s.io/descheduler/pkg/descheduler.RunDeschedulerStrategies.func1()
	/go/src/sigs.k8s.io/descheduler/pkg/descheduler/descheduler.go:145 +0x26c
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x4000707ad0)
	/go/src/sigs.k8s.io/descheduler/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x64
k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0x4000707ad0, 0x19c1ff0, 0x4000454210, 0x1, 0x4000028420)
	/go/src/sigs.k8s.io/descheduler/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0x74
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0x4000707ad0, 0x0, 0x0, 0x1, 0x4000028420)
	/go/src/sigs.k8s.io/descheduler/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x88
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go/src/sigs.k8s.io/descheduler/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90
sigs.k8s.io/descheduler/pkg/descheduler.RunDeschedulerStrategies(0x1a00468, 0x400019c020, 0x400046c2c0, 0x40001a6140, 0x40003be487, 0x9, 0x4000028420, 0x4000092c00, 0x40000ee460)
	/go/src/sigs.k8s.io/descheduler/pkg/descheduler/descheduler.go:117 +0x534
sigs.k8s.io/descheduler/pkg/descheduler.Run(0x400046c2c0, 0x19c2050, 0x400021c0e0)
	/go/src/sigs.k8s.io/descheduler/pkg/descheduler/descheduler.go:64 +0x154
sigs.k8s.io/descheduler/cmd/descheduler/app.Run(...)
	/go/src/sigs.k8s.io/descheduler/cmd/descheduler/app/server.go:93
sigs.k8s.io/descheduler/cmd/descheduler/app.NewDeschedulerCommand.func1(0x40003ba280, 0x400012b7c0, 0x0, 0x4)
	/go/src/sigs.k8s.io/descheduler/cmd/descheduler/app/server.go:78 +0xb8
github.com/spf13/cobra.(*Command).execute(0x40003ba280, 0x40001a6010, 0x4, 0x4, 0x40003ba280, 0x40001a6010)
	/go/src/sigs.k8s.io/descheduler/vendor/github.com/spf13/cobra/command.go:856 +0x1d0
github.com/spf13/cobra.(*Command).ExecuteC(0x40003ba280, 0x400063ff48, 0x1, 0x1)
	/go/src/sigs.k8s.io/descheduler/vendor/github.com/spf13/cobra/command.go:960 +0x268
github.com/spf13/cobra.(*Command).Execute(...)
	/go/src/sigs.k8s.io/descheduler/vendor/github.com/spf13/cobra/command.go:897
main.main()
	/go/src/sigs.k8s.io/descheduler/cmd/descheduler/descheduler.go:34 +0xa0

```

--- a/pkg/descheduler/strategies/failedpods.go
+++ b/pkg/descheduler/strategies/failedpods.go
@@ -90,7 +90,9 @@ func validateAndParseRemoveFailedPodsParams(
 	params *api.StrategyParameters,
 ) (*validatedFailedPodsStrategyParams, error) {
 	if params == nil {
-		return &validatedFailedPodsStrategyParams{}, nil
+		return &validatedFailedPodsStrategyParams{
+			ValidatedStrategyParams: validation.DefaultValidatedStrategyParams(),
+		}, nil
 	}
 
 	strategyParams, err := validation.ValidateAndParseStrategyParams(ctx, client, params)

--- a/pkg/descheduler/strategies/failedpods.go
+++ b/pkg/descheduler/strategies/failedpods.go
@@ -19,7 +19,7 @@ import (
 
 // validatedFailedPodsStrategyParams contains validated strategy parameters
 type validatedFailedPodsStrategyParams struct {
-	*validation.ValidatedStrategyParams
+	validation.ValidatedStrategyParams
 	includingInitContainers bool
 	reasons                 sets.String
 	excludeOwnerKinds       sets.String
@@ -46,9 +46,15 @@ func RemoveFailedPods(
 		evictions.WithLabelSelector(strategyParams.LabelSelector),
 	)
 
+	var labelSelector *metav1.LabelSelector
+	if strategy.Params != nil {
+		labelSelector = strategy.Params.LabelSelector
+	}
+
 	for _, node := range nodes {
 		klog.V(1).InfoS("Processing node", "node", klog.KObj(node))
 		fieldSelectorString := "spec.nodeName=" + node.Name + ",status.phase=" + string(v1.PodFailed)
+
 		pods, err := podutil.ListPodsOnANodeWithFieldSelector(
 			ctx,
 			client,
@@ -57,7 +63,7 @@ func RemoveFailedPods(
 			podutil.WithFilter(evictable.IsEvictable),
 			podutil.WithNamespaces(strategyParams.IncludedNamespaces.UnsortedList()),
 			podutil.WithoutNamespaces(strategyParams.ExcludedNamespaces.UnsortedList()),
-			podutil.WithLabelSelector(strategy.Params.LabelSelector),
+			podutil.WithLabelSelector(labelSelector),
 		)
 		if err != nil {
 			klog.ErrorS(err, "Error listing a nodes failed pods", "node", klog.KObj(node))
@@ -103,7 +109,7 @@ func validateAndParseRemoveFailedPodsParams(
 	}
 
 	return &validatedFailedPodsStrategyParams{
-		ValidatedStrategyParams: strategyParams,
+		ValidatedStrategyParams: *strategyParams,
 		includingInitContainers: includingInitContainers,
 		reasons:                 reasons,
 		excludeOwnerKinds:       excludeOwnerKinds,

--- a/pkg/descheduler/strategies/failedpods_test.go
+++ b/pkg/descheduler/strategies/failedpods_test.go
@@ -45,6 +45,13 @@ func TestRemoveFailedPods(t *testing.T) {
 		pods                    []v1.Pod
 	}{
 		{
+			description:             "default empty strategy, 0 failures, 0 evictions",
+			strategy:                api.DeschedulerStrategy{},
+			nodes:                   []*v1.Node{test.BuildTestNode("node1", 2000, 3000, 10, nil)},
+			expectedEvictedPodCount: 0,
+			pods:                    []v1.Pod{}, // no pods come back with field selector phase=Failed
+		},
+		{
 			description:             "0 failures, 0 evictions",
 			strategy:                createStrategy(true, false, nil, nil, nil, false),
 			nodes:                   []*v1.Node{test.BuildTestNode("node1", 2000, 3000, 10, nil)},
@@ -225,6 +232,7 @@ func TestValidRemoveFailedPodsParams(t *testing.T) {
 		params *api.StrategyParameters
 	}{
 		{name: "validate nil params", params: nil},
+		{name: "validate empty params", params: &api.StrategyParameters{}},
 		{name: "validate reasons params", params: &api.StrategyParameters{FailedPods: &api.FailedPods{
 			Reasons: []string{"CreateContainerConfigError"},
 		}}},

--- a/pkg/descheduler/strategies/validation/strategyparams.go
+++ b/pkg/descheduler/strategies/validation/strategyparams.go
@@ -22,20 +22,22 @@ type ValidatedStrategyParams struct {
 	NodeFit            bool
 }
 
+func DefaultValidatedStrategyParams() ValidatedStrategyParams {
+	return ValidatedStrategyParams{ThresholdPriority: utils.SystemCriticalPriority}
+}
+
 func ValidateAndParseStrategyParams(
 	ctx context.Context,
 	client clientset.Interface,
 	params *api.StrategyParameters,
 ) (*ValidatedStrategyParams, error) {
-	var includedNamespaces, excludedNamespaces sets.String
 	if params == nil {
-		return &ValidatedStrategyParams{
-			IncludedNamespaces: includedNamespaces,
-			ExcludedNamespaces: excludedNamespaces,
-		}, nil
+		defaultValidatedStrategyParams := DefaultValidatedStrategyParams()
+		return &defaultValidatedStrategyParams, nil
 	}
 
 	// At most one of include/exclude can be set
+	var includedNamespaces, excludedNamespaces sets.String
 	if params.Namespaces != nil && len(params.Namespaces.Include) > 0 && len(params.Namespaces.Exclude) > 0 {
 		return nil, fmt.Errorf("only one of Include/Exclude namespaces can be set")
 	}

--- a/test/e2e/e2e_failedpods_test.go
+++ b/test/e2e/e2e_failedpods_test.go
@@ -1,0 +1,154 @@
+package e2e
+
+import (
+	"context"
+	batchv1 "k8s.io/api/batch/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+	clientset "k8s.io/client-go/kubernetes"
+	"strings"
+	"testing"
+	"time"
+
+	deschedulerapi "sigs.k8s.io/descheduler/pkg/api"
+	"sigs.k8s.io/descheduler/pkg/descheduler/strategies"
+)
+
+var oneHourPodLifetimeSeconds uint = 3600
+
+func TestFailedPods(t *testing.T) {
+	ctx := context.Background()
+	clientSet, _, stopCh := initializeClient(t)
+	defer close(stopCh)
+	nodeList, err := clientSet.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		t.Errorf("Error listing node with %v", err)
+	}
+	nodes, _ := splitNodesAndWorkerNodes(nodeList.Items)
+	t.Log("Creating testing namespace")
+	testNamespace := &v1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "e2e-" + strings.ToLower(t.Name())}}
+	if _, err := clientSet.CoreV1().Namespaces().Create(ctx, testNamespace, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Unable to create ns %v", testNamespace.Name)
+	}
+	defer clientSet.CoreV1().Namespaces().Delete(ctx, testNamespace.Name, metav1.DeleteOptions{})
+	testCases := map[string]struct {
+		expectedEvictedCount int
+		strategyParams       *deschedulerapi.StrategyParameters
+	}{
+		"test-failed-pods-nil-strategy": {
+			expectedEvictedCount: 1,
+			strategyParams:       nil,
+		},
+		"test-failed-pods-default-strategy": {
+			expectedEvictedCount: 1,
+			strategyParams:       &deschedulerapi.StrategyParameters{},
+		},
+		"test-failed-pods-default-failed-pods": {
+			expectedEvictedCount: 1,
+			strategyParams: &deschedulerapi.StrategyParameters{
+				FailedPods: &deschedulerapi.FailedPods{},
+			},
+		},
+		"test-failed-pods-reason-unmatched": {
+			expectedEvictedCount: 0,
+			strategyParams: &deschedulerapi.StrategyParameters{
+				FailedPods: &deschedulerapi.FailedPods{Reasons: []string{"ReasonDoesNotMatch"}},
+			},
+		},
+		"test-failed-pods-min-age-unmet": {
+			expectedEvictedCount: 0,
+			strategyParams: &deschedulerapi.StrategyParameters{
+				FailedPods: &deschedulerapi.FailedPods{MinPodLifetimeSeconds: &oneHourPodLifetimeSeconds},
+			},
+		},
+		"test-failed-pods-exclude-job-kind": {
+			expectedEvictedCount: 0,
+			strategyParams: &deschedulerapi.StrategyParameters{
+				FailedPods: &deschedulerapi.FailedPods{ExcludeOwnerKinds: []string{"Job"}},
+			},
+		},
+	}
+	for name, tc := range testCases {
+		t.Run(name, func(t *testing.T) {
+			job := initFailedJob(name, testNamespace.Namespace)
+			t.Logf("Creating job %s in %s namespace", job.Name, job.Namespace)
+			jobClient := clientSet.BatchV1().Jobs(testNamespace.Name)
+			if _, err := jobClient.Create(ctx, job, metav1.CreateOptions{}); err != nil {
+				t.Fatalf("Error creating Job %s: %v", name, err)
+			}
+			deletePropagationPolicy := metav1.DeletePropagationForeground
+			defer jobClient.Delete(ctx, job.Name, metav1.DeleteOptions{PropagationPolicy: &deletePropagationPolicy})
+			waitForJobPodPhase(ctx, t, clientSet, job, v1.PodFailed)
+
+			podEvictor := initPodEvictorOrFail(t, clientSet, nodes)
+
+			t.Logf("Running RemoveFailedPods strategy for %s", name)
+			strategies.RemoveFailedPods(
+				ctx,
+				clientSet,
+				deschedulerapi.DeschedulerStrategy{
+					Enabled: true,
+					Params:  tc.strategyParams,
+				},
+				nodes,
+				podEvictor,
+			)
+			t.Logf("Finished RemoveFailedPods strategy for %s", name)
+
+			if actualEvictedCount := podEvictor.TotalEvicted(); actualEvictedCount == tc.expectedEvictedCount {
+				t.Logf("Total of %d Pods were evicted for %s", actualEvictedCount, name)
+			} else {
+				t.Errorf("Unexpected number of pods have been evicted, got %v, expected %v", actualEvictedCount, tc.expectedEvictedCount)
+			}
+		})
+	}
+}
+
+func initFailedJob(name, namespace string) *batchv1.Job {
+	podSpec := MakePodSpec("", nil)
+	podSpec.Containers[0].Command = []string{"/bin/false"}
+	podSpec.RestartPolicy = v1.RestartPolicyNever
+	labelsSet := labels.Set{"test": name, "name": name}
+	jobBackoffLimit := int32(0)
+	return &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:    labelsSet,
+			Name:      name,
+			Namespace: namespace,
+		},
+		Spec: batchv1.JobSpec{
+			Template: v1.PodTemplateSpec{
+				Spec:       podSpec,
+				ObjectMeta: metav1.ObjectMeta{Labels: labelsSet},
+			},
+			BackoffLimit: &jobBackoffLimit,
+		},
+	}
+}
+
+func waitForJobPodPhase(ctx context.Context, t *testing.T, clientSet clientset.Interface, job *batchv1.Job, phase v1.PodPhase) {
+	podClient := clientSet.CoreV1().Pods(job.Namespace)
+	if err := wait.PollImmediate(5*time.Second, 30*time.Second, func() (bool, error) {
+		t.Log(labels.FormatLabels(job.Labels))
+		if podList, err := podClient.List(ctx, metav1.ListOptions{LabelSelector: labels.FormatLabels(job.Labels)}); err != nil {
+			return false, err
+		} else {
+			if len(podList.Items) == 0 {
+				t.Logf("Job controller has not created Pod for job %s yet", job.Name)
+				return false, nil
+			}
+			for _, pod := range podList.Items {
+				if pod.Status.Phase != phase {
+					t.Logf("Pod %v not in %s phase yet, is %v instead", pod.Name, phase, pod.Status.Phase)
+					return false, nil
+				}
+			}
+			t.Logf("Job %v Pod is in %s phase now", job.Name, phase)
+			return true, nil
+		}
+	}); err != nil {
+		t.Fatalf("Error waiting for pods in %s phase: %v", phase, err)
+	}
+}

--- a/test/e2e/e2e_test.go
+++ b/test/e2e/e2e_test.go
@@ -290,20 +290,7 @@ func TestLowNodeUtilization(t *testing.T) {
 	waitForRCPodsRunning(ctx, t, clientSet, rc)
 
 	// Run LowNodeUtilization strategy
-	evictionPolicyGroupVersion, err := eutils.SupportEviction(clientSet)
-	if err != nil || len(evictionPolicyGroupVersion) == 0 {
-		t.Fatalf("%v", err)
-	}
-	podEvictor := evictions.NewPodEvictor(
-		clientSet,
-		evictionPolicyGroupVersion,
-		false,
-		0,
-		nodes,
-		true,
-		false,
-		false,
-	)
+	podEvictor := initPodEvictorOrFail(t, clientSet, nodes)
 
 	podsOnMosttUtilizedNode, err := podutil.ListPodsOnANode(ctx, clientSet, workerNodes[0], podutil.WithFilter(podEvictor.Evictable().IsEvictable))
 	if err != nil {
@@ -1292,4 +1279,21 @@ func splitNodesAndWorkerNodes(nodes []v1.Node) ([]*v1.Node, []*v1.Node) {
 		}
 	}
 	return allNodes, workerNodes
+}
+
+func initPodEvictorOrFail(t *testing.T, clientSet clientset.Interface, nodes []*v1.Node) *evictions.PodEvictor {
+	evictionPolicyGroupVersion, err := eutils.SupportEviction(clientSet)
+	if err != nil || len(evictionPolicyGroupVersion) == 0 {
+		t.Fatalf("Error creating eviction policy group: %v", err)
+	}
+	return evictions.NewPodEvictor(
+		clientSet,
+		evictionPolicyGroupVersion,
+		false,
+		0,
+		nodes,
+		true,
+		false,
+		false,
+	)
 }

--- a/test/e2e/e2e_topologyspreadconstraint_test.go
+++ b/test/e2e/e2e_topologyspreadconstraint_test.go
@@ -11,8 +11,6 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
 	deschedulerapi "sigs.k8s.io/descheduler/pkg/api"
-	"sigs.k8s.io/descheduler/pkg/descheduler/evictions"
-	eutils "sigs.k8s.io/descheduler/pkg/descheduler/evictions/utils"
 	"sigs.k8s.io/descheduler/pkg/descheduler/strategies"
 )
 
@@ -79,22 +77,9 @@ func TestTopologySpreadConstraint(t *testing.T) {
 			defer deleteRC(ctx, t, clientSet, violatorRc)
 			waitForRCPodsRunning(ctx, t, clientSet, violatorRc)
 
-			// Run TopologySpreadConstraint strategy
-			evictionPolicyGroupVersion, err := eutils.SupportEviction(clientSet)
-			if err != nil || len(evictionPolicyGroupVersion) == 0 {
-				t.Fatalf("Error creating eviction policy group for %s: %v", name, err)
-			}
-			podEvictor := evictions.NewPodEvictor(
-				clientSet,
-				evictionPolicyGroupVersion,
-				false,
-				0,
-				nodes,
-				true,
-				false,
-				false,
-			)
+			podEvictor := initPodEvictorOrFail(t, clientSet, nodes)
 
+			// Run TopologySpreadConstraint strategy
 			t.Logf("Running RemovePodsViolatingTopologySpreadConstraint strategy for %s", name)
 			strategies.RemovePodsViolatingTopologySpreadConstraint(
 				ctx,

--- a/test/e2e/e2e_topologyspreadconstraint_test.go
+++ b/test/e2e/e2e_topologyspreadconstraint_test.go
@@ -119,6 +119,9 @@ func TestTopologySpreadConstraint(t *testing.T) {
 				t.Fatalf("Pods were not evicted for %s TopologySpreadConstraint", name)
 			}
 
+			// Ensure recently evicted Pod are rescheduled and running before asserting for a balanced topology spread
+			waitForRCPodsRunning(ctx, t, clientSet, rc)
+
 			pods, err := clientSet.CoreV1().Pods(testNamespace.Name).List(ctx, metav1.ListOptions{LabelSelector: fmt.Sprintf("%s=%s", tc.labelKey, tc.labelValue)})
 			if err != nil {
 				t.Errorf("Error listing pods for %s: %v", name, err)

